src/providers/twitch/IrcMessageHandler.cpp

```diff
@@ -507,129 +297,370 @@ void IrcMessageHandler::populateReply(
+        if (ptr)
         {
-            builder->flags.set(MessageFlag::Subscription);
-            builder->flags.unset(MessageFlag::Highlighted);
+            builtMessages.emplace_back(std::move(ptr));
         }
-        auto msg = builder.build();
+    }
 
-        IrcMessageHandler::setSimilarityFlags(msg, chan);
+    return builtMessages;
+}
 
-        if (!msg->flags.has(MessageFlag::Similar) ||
-            (!getSettings()->hideSimilar &&
-             getSettings()->shownSimilarTriggerHighlights))
+/**
+ * Parse a single IRC message into 0 or more Chatterino messages
+ **/
+std::vector<MessagePtr> parseMessage(Channel *channel,
+                                     Communi::IrcMessage *message)
+{
+    assert(channel != nullptr);
+    assert(message != nullptr);
+
+    std::vector<MessagePtr> builtMessages;
+
+    auto command = message->command();
+
+    if (command == "PRIVMSG")
+    {
+        return parsePrivMessage(
+            channel, dynamic_cast<Communi::IrcPrivateMessage *>(message));
+    }
+
+    if (command == "USERNOTICE")
+    {
+        return parseUserNoticeMessage(channel, message);
+    }
+
+    if (command == "NOTICE")
+    {
+        return parseNoticeMessage(
+            dynamic_cast<Communi::IrcNoticeMessage *>(message));
+    }
+
+    return builtMessages;
+}
+
+}  // namespace
+
+namespace chatterino {
+
+using namespace literals;
+
+IrcMessageHandler &IrcMessageHandler::instance()
+{
+    static IrcMessageHandler instance;
+    return instance;
+}
+
+std::vector<MessagePtr> IrcMessageHandler::parseMessageWithReply(
+    Channel *channel, Communi::IrcMessage *message,
+    std::vector<MessagePtr> &otherLoaded)
+{
+    std::vector<MessagePtr> builtMessages;
+
+    auto command = message->command();
+
+    if (command == "PRIVMSG")
+    {
+        auto *privMsg = dynamic_cast<Communi::IrcPrivateMessage *>(message);
+        auto *tc = dynamic_cast<TwitchChannel *>(channel);
+        if (!tc)
         {
-            builder.triggerHighlights();
+            return parsePrivMessage(channel, privMsg);
         }
 
-        const auto highlighted = msg->flags.has(MessageFlag::Highlighted);
-        const auto showInMentions = msg->flags.has(MessageFlag::ShowInMentions);
+        QString content = privMsg->content();
+        int messageOffset = stripLeadingReplyMention(privMsg->tags(), content);
+        MessageParseArgs args;
+        TwitchMessageBuilder builder(channel, message, args, content,
+                                     privMsg->isAction());
+        builder.setMessageOffset(messageOffset);
 
-        if (highlighted && showInMentions)
+        populateReply(tc, message, otherLoaded, builder);
+
+        if (!builder.isIgnored())
         {
-            server.mentionsChannel->addMessage(msg);
+            builtMessages.emplace_back(builder.build());
+            builder.triggerHighlights();
         }
+    }
+    else if (command == "USERNOTICE")
```

Comment:
If there was an early return here, the `else if`s could be turned into `if`s.
```suggestion
        }
        return builtMessages;
    }

    if (command == "USERNOTICE")
```

Reply:
Resolved in [`c91f947` (#4927)](https://github.com/Chatterino/chatterino2/pull/4927/commits/c91f94744fc0d26c56ee23d4049b9d20eba9cd30), with bonus `u"xd"_s` fixes in [`441e50e` (#4927)](https://github.com/Chatterino/chatterino2/pull/4927/commits/441e50e300de8272109cec4a73584c4682cdc942)